### ISSUE TITLE
6.0-beta1 stuck with "Waiting for React to connect..."

### ISSUE THREAD

**Issue by pocesar** [jhen0409/react-native-debugger#48]
The Debug JS works, the packager is working, but the React / Redux tools doesn't load, it's stuck in "Waiting for React to connect" even though it's already being able to log and replay redux actions, and I can set breakpoints and it does work. 

![image](https://cloud.githubusercontent.com/assets/461084/23221329/8a54672a-f903-11e6-8ea9-f0fdb76ef040.png)

On windows 10 x64, node 7.50 and react-native 0.41.2 / react 15.4.2

Using a real device through Wifi (using adb tcpip 5555 / adb shell ip route / adb connect 192.168.0.109:5555

**Comment by jhen0409** [comment]
> Using a real device through Wifi (using adb tcpip 5555 / adb shell ip route / adb connect 192.168.0.109:5555

You need to change [hostname](https://github.com/facebook/react-native/blob/bf2b435322e89d0aeee8792b1c6e04656c2719a0/Libraries/Core/Devtools/setupDevtools.js#L20) to your machine IP, have you tried that?

It already in [readme](https://github.com/jhen0409/react-native-debugger#react-inspector), although the link is wrong for now. 😬

**Comment by pocesar** [comment]
ah, that's the reason why I couldn't find the setting, the README is indeed pointing to something totally unrelated...

**Comment by pocesar** [comment]
changed the like to: 

```js
  if (Platform.OS === 'android' && NativeModules.AndroidConstants) {
    hostname = '192.168.0.107';//NativeModules.AndroidConstants.ServerHost.split(':')[0];
  }
```

but now it shows "Connecting to React…" and stuck

**Comment by jhen0409** [comment]
@pocesar I think [this section](https://github.com/jhen0409/react-native-debugger#android-react-inspector-get-stuck-at-connecting-to-react-45) will help you, this is related to recently `react-devtools` changes. Maybe it's `requestIdleCallback` bug in debug mode of RN.

**Comment by pocesar** [comment]
works now, but those modifications are non-portable. isn't there a better way? there's no problem with the time (both device and dev computer uses NTP pool, plus my router as well), at least the part that needs to edit setupDevTools.js manually, is there anyway to set that variable programaticaly?

**Comment by jhen0409** [comment]
> at least the part that needs to edit setupDevTools.js manually, is there anyway to set that variable programaticaly?

Currently you can use `connectToDevTools` of [`react-devtools-core`](https://github.com/facebook/react-devtools/tree/master/packages/react-devtools-core#requirereact-devtools-corestandalone) instead of built-in `setupDevtools`.

```js
// Make sure it before imported at `import { ... } from 'react-native'`
import { connectToDevTools } from 'react-devtools-core';
import { ... } from 'react-native'

connectToDevTools({
  host: '<yourIP>',
  resolveRNStyle: require('flattenStyle')
})
```

**Comment by pocesar** [comment]
it's giving this

```
[React DevTools] Connected
F:\Users\Paulo\Desktop\RNDebugger\resources\app.asar\node_modules\react-devtools-core\build\standal…:97 [React DevTools] Only one connection allowed at a time. Closing the previous connection
log.warn @ F:\Users\Paulo\Desktop\RNDebugger\resources\app.asar\node_modules\react-devtools-core\build\standal…:97
(anonymous) @ F:\Users\Paulo\Desktop\RNDebugger\resources\app.asar\node_modules\react-devtools-core\build\standal…:56
emitOne @ events.js:96
emit @ events.js:188
handleUpgrade @ F:\Users\Paulo\Desktop\RNDebugger\resources\app.asar\node_modules\ws\lib\WebSocketServer.js:87
completeUpgrade @ F:\Users\Paulo\Desktop\RNDebugger\resources\app.asar\node_modules\ws\lib\WebSocketServer.js:272
handleUpgrade @ F:\Users\Paulo\Desktop\RNDebugger\resources\app.asar\node_modules\ws\lib\WebSocketServer.js:199
WebSocketServer._ultron.on @ F:\Users\Paulo\Desktop\RNDebugger\resources\app.asar\node_modules\ws\lib\WebSocketServer.js:85
emitThree @ events.js:116
emit @ events.js:194
onParserExecuteCommon @ _http_server.js:459
onParserExecute @ _http_server.js:418
F:\Users\Paulo\Desktop\RNDebugger\resources\app.asar\node_modules\react-devtools-core\build\standal…:93 [React DevTools] Connected
F:\Users\Paulo\Desktop\RNDebugger\resources\app.asar\node_modules\react-devtools-core\build\standal…:93 [React DevTools] Connection to RN closed
F:\Users\Paulo\Desktop\RNDebugger\resources\app.asar\node_modules\react-devtools-core\build\standal…:93 [React DevTools] Connected
F:\Users\Paulo\Desktop\RNDebugger\resources\app.asar\node_modules\react-devtools-core\build\standal…:12409 failed to connect
```

**Comment by jhen0409** [comment]
@pocesar maybe you need to revert `setupDevtools` change, or just set `window.__REACT_DEVTOOLS_PORT__` with a different port.

**Comment by pocesar** [comment]
yeah, it doesn't work, I guess I'll need to stick with manually patching the setupDevtools file

**Comment by jhen0409** [comment]
hmm it should work, and currently the usage have been replaced old `setupDevtools` on [master of RN](https://github.com/facebook/react-native/blob/master/Libraries/Core/Devtools/setupDevtools.js).

Make sure it before imported at `import { ... } from 'react-native'` (can be written at the top of `index.android.js`), otherwise I can try it again for Windows.

**Comment by MasterAM** [comment]
I have experienced similar symptoms. 
In my case, it suddenly started taking very long for the component view to show ("Connecting to React..."), but the console was functioning as normal.

Then, when I was initiating an action that triggered substantial animation, the UI would freeze (physical device) and the dev tools would be disconnected for up to a couple of minutes.

At some point, after attempting to isolate the issue and having everything working well on another physical device and on a different machine, I noticed that fixing a **clock drift** made everything work again (by changing the device time manually, as the mobile network time was way off).

I did not get a warning about the drift, and it drifted about 1 minute (perhaps less) from my server time.
